### DISCUSSION TITLE
Customize log rotating behavior

### DISCUSSION THREAD

**Discussion by lyudmilalala** [seaweedfs/seaweedfs#5057] [Q&A]
1. How is seaweedfs log rotation implemented? How can I find the default setting for seaweedfs log rotation?
2. Can someone show me an example of customizing log rotation period of file size using `weed [logging_options] [command]`? I am confuse about the exactly configs I should input.
3. If I have a weed master server, a weed volume server, and a weed filer server on a same node, it seems that filer logs will not be write into the default log files in `/tmp`. (I compare the stdout of each server with the log file in  `/tmp`.) Am I correct? How can I persist my filer logs? I would implement high availability across nodes so I did not use `weed server`.

As files as the following are generated in `/tmp`, I think there is a default log rotation action.

```
-rw-r--r--  1 root  root   12K Nov 28 10:40 weed.master.root.log.ERROR.20231128-095950.3693475
-rw-r--r--  1 root  root  2.5K Nov 28 12:34 weed.master.root.log.ERROR.20231128-123431.3856682
-rw-r--r--  1 root  root  2.5K Nov 28 09:47 weed.master.root.log.FATAL.20231128-094704.3681746
-rw-r--r--  1 root  root  2.5K Nov 28 09:47 weed.master.root.log.FATAL.20231128-094753.3682749
-rw-r--r--  1 root  root  2.6K Nov 28 12:34 weed.master.root.log.INFO.20231128-123431.3856682
-rw-r--r--  1 root  root   23K Nov 28 14:04 weed.master.root.log.INFO.20231128-123457.3857207
-rw-r--r--  1 root  root   18K Nov 28 10:40 weed.master.root.log.WARNING.20231128-095813.3693475
-rw-r--r--  1 root  root   443 Nov 28 10:40 weed.master.root.log.WARNING.20231128-104004.3731221
```

I have a look at the [glog implementation](https://github.com/seaweedfs/seaweedfs/blob/master/weed/glog/glog.go), but still don't know how to set up the function I want.

**Comment by chrislusf** [comment]
glog does not have log rotation feature. There might be some other fork having this feature.

**Reply by lyudmilalala** [reply]
So how could I rotate the log by size or time as I expect?

**Comment by lyudmilalala** [comment]
Hi all,

After a few research I find the max size and max count of log files are defined[here](https://github.com/seaweedfs/seaweedfs/blob/master/weed/glog/glog_file.go#L35) and determined [here](https://github.com/seaweedfs/seaweedfs/blob/master/weed/weed.go#L49).

If you want to rotate by time, or do more customization, you should probably modify [line 813 to 878 in the glog.go](https://github.com/seaweedfs/seaweedfs/blob/master/weed/glog/glog.go#L817).

Following is a small demo of how it works.

---

tart a docker container for build the binary weed. Mount the directory with source code on it.

```
docker run -itd --name go-builder  -p 9333:9333 -p 8888:8888 -p 8080:8080 -p 8333:8333  -v home/mila/seaweedfs/build/src:/opt/weed_src golang:1.21
```

Get into the docker container

```
docker exec -it go-builder /bin/bash
```

Change `seaweedfs-3.57/weed/weed.go` line 49-50.

```
glog.MaxSize  = 1024 * 125
glog.MaxFileCount = 3
```

Compile the source code 

```
cd /opt/weed_src/seaweedfs-3.57/weed && make install
```

Now the weed binary has been generated at `/go/bin/weed`. Move it to `$PATH`.

```
cp /go/bin/weed /usr/local/bin/
```

Run a simple server

```shell
weed -logdir=/opt/log/seaweedfs server -dir=/opt/data -s3 -master.volumeSizeLimitMB=256 -volume.max=16
```

Use a script as the following to request it to generate some log

```python
import boto3
import time

SEAWEED_S3_SERVER = 'http://192.168.8.171:8333'
s3 = boto3.client('s3',endpoint_url=SEAWEED_S3_SERVER,aws_access_key_id='123',aws_secret_access_key='456')
bucket_name = 'demo'

def create_bucket_if_not_exist(bucket_name):
    # list buckets
    response = s3.list_buckets()
    buckets = [bucket['Name'] for bucket in response['Buckets']]
    if bucket_name not in buckets:
        # Create the bucket if it doesn't exist
        s3.create_bucket(Bucket=bucket_name)
        print(f"Bucket '{bucket_name}' created")
    else:
        print(f"Bucket '{bucket_name}' already exists")

for i in range(20):
    try:
        create_bucket_if_not_exist(bucket_name)
        # Upload a file to the bucket
        with open('data/longText.txt', 'rb') as data:
            s3.upload_fileobj(data, bucket_name, 'longText.txt')
    except Exception as e:
        print(f"Error uploading file: {e}")
        exit(1)

    time.sleep(2)
    response = s3.delete_bucket(Bucket=bucket_name)
```

Log at `/opt/log/seaweedfs` should rotate properly.

```
drwxrwxrwx 1 root root 4.0K Dec 11 06:45 .
drwxr-xr-x 3 root root 4.0K Dec  7 07:25 ..
-rw-r--r-- 1 root root  23K Dec 11 06:46 weed.057c35efd900.root.log.ERROR.20231211-063416.10715
-rw-r--r-- 1 root root 125K Dec 11 06:37 weed.057c35efd900.root.log.INFO.20231211-063503.10715
-rw-r--r-- 1 root root 125K Dec 11 06:45 weed.057c35efd900.root.log.INFO.20231211-063753.10715
-rw-r--r-- 1 root root  20K Dec 11 07:54 weed.057c35efd900.root.log.INFO.20231211-064557.10715
-rw-r--r-- 1 root root  23K Dec 11 07:54 weed.057c35efd900.root.log.WARNING.20231211-062727.10715
lrwxrwxrwx 1 root root   54 Dec 11 06:34 weed.ERROR -> weed.057c35efd900.root.log.ERROR.20231211-063416.10715
lrwxrwxrwx 1 root root   53 Dec 11 06:45 weed.INFO -> weed.057c35efd900.root.log.INFO.20231211-064557.10715
lrwxrwxrwx 1 root root   56 Dec 11 06:27 weed.WARNING -> weed.057c35efd900.root.log.WARNING.20231211-062727.10715
```
